core/src/transport/upgrade.rs

```diff
@@ -99,16 +99,15 @@ where
     ) -> Authenticated<AndThen<T, impl FnOnce(C, ConnectedPoint) -> Authenticate<C, U> + Clone>>
     where
         T: Transport<Output = C>,
-        C: AsyncRead + AsyncWrite + Unpin,
+        C: AsyncRead + AsyncWrite + Unpin + 'static,
         D: AsyncRead + AsyncWrite + Unpin,
-        U: InboundConnectionUpgrade<Negotiated<C>, Output = (PeerId, D), Error = E>,
-        U: OutboundConnectionUpgrade<Negotiated<C>, Output = (PeerId, D), Error = E> + Clone,
+        U: SecurityUpgrade<Negotiated<C>, Output = (PeerId, D), Error = E> + Clone + 'static,
         E: Error + 'static,
     {
         let version = self.version;
         Authenticated(Builder::new(
             self.inner.and_then(move |conn, endpoint| Authenticate {
-                inner: upgrade::apply(conn, upgrade, endpoint, version),
+                inner: upgrade::secure(conn, upgrade, endpoint, version).boxed_local(),
```

Comment:
Why `boxed_local`?

Reply:
Because the function parameters are not `Send`.

Reply:
I think effectively, they will be so we can just add the necessary constraints here. We need the returned future to be `Send` otherwise we cannot spawn it onto the executor.

Reply:
That was my impression too, but I didn't want to add new restrictions to the types from the beginning.
Following this approach we also need some associated types to be `Send`.